File: internal/drivers/smtp/smtp.go
// Package smtp is the raw SMTP driver.
//
// SMTP_URL forms:
//
//	smtp://user@host:587?starttls=true
//	smtps://user@host:465
//
// The password comes separately (Docker secret smtp_password).
package smtp

import (
	"crypto/tls"
	"net"
	"net/smtp"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	smtpPort    = 25
	smtpsPort   = 465
	schemeSSL   = "smtps"
	dialTimeout = 20 * time.Second
)

// Target is a parsed SMTP_URL.
type Target struct {
	Host     string
	Port     int
	User     string
	UseSSL   bool
	StartTLS bool
}

// Parse reads an SMTP_URL.
func Parse(raw string) (Target, error) {
	u, err := url.Parse(raw)
	if err != nil {
		return Target{}, err
	}

	t := Target{Host: u.Hostname(), UseSSL: u.Scheme == schemeSSL}
	if t.Host == "" {
		t.Host = "localhost"
	}

	t.Port = smtpPort
	if t.UseSSL {
		t.Port = smtpsPort
	}
	if p := u.Port(); p != "" {
		if t.Port, err = strconv.Atoi(p); err != nil {
			return Target{}, err
		}
	}

	if u.User != nil {
		t.User = u.User.Username()
	}
	t.StartTLS = strings.EqualFold(u.Query().Get("starttls"), "true")
	return t, nil
}

// Send delivers one prepared RFC 5322 message.
func Send(t Target, password, from string, to []string, message []byte) error {
	addr := net.JoinHostPort(t.Host, strconv.Itoa(t.Port))
	tlsCfg := &tls.Config{ServerName: t.Host, MinVersion: tls.VersionTLS12}

	conn, err := dial(t, addr, tlsCfg)
	if err != nil {
		return err
	}

	client, err := smtp.NewClient(conn, t.Host)
	if err != nil {
		conn.Close()
		return err
	}
	defer client.Close()

	if t.StartTLS {
		if err := client.StartTLS(tlsCfg); err != nil {
			return err
		}
	}
	if t.User != "" && password != "" {
		if err := client.Auth(smtp.PlainAuth("", t.User, password, t.Host)); err != nil {
			return err
		}
	}

	if err := client.Mail(from); err != nil {
		return err
	}
	for _, rcpt := range to {
		if err := client.Rcpt(rcpt); err != nil {
			return err
		}
	}

	w, err := client.Data()
	if err != nil {
		return err
	}
	if _, err := w.Write(message); err != nil {
		return err
	}
	if err := w.Close(); err != nil {
		return err
	}
	return client.Quit()
}

func dial(t Target, addr string, tlsCfg *tls.Config) (net.Conn, error) {
	dialer := &net.Dialer{Timeout: dialTimeout}
	if t.UseSSL {
		return tls.DialWithDialer(dialer, "tcp", addr, tlsCfg)
	}
	return dialer.Dial("tcp", addr)
}

File: app/drivers/smtp.py
"""Raw SMTP delivery.

SMTP_URL forms:
    smtp://user@host:587?starttls=true
    smtps://user@host:465
Password comes separately (Docker secret smtp_password).
"""

import smtplib
import ssl
from dataclasses import dataclass
from email.message import EmailMessage
from urllib.parse import parse_qs, unquote, urlparse

SMTP_PORT = 25
SMTPS_PORT = 465
TIMEOUT_S = 20
SCHEME_SSL = "smtps"


@dataclass(frozen=True)
class SmtpTarget:
    host: str
    port: int
    user: str | None
    use_ssl: bool
    starttls: bool


def parse(url: str) -> SmtpTarget:
    parts = urlparse(url)
    query = parse_qs(parts.query)
    use_ssl = parts.scheme == SCHEME_SSL
    return SmtpTarget(
        host=parts.hostname or "localhost",
        port=parts.port or (SMTPS_PORT if use_ssl else SMTP_PORT),
        user=unquote(parts.username) if parts.username else None,
        use_ssl=use_ssl,
        starttls=query.get("starttls", ["false"])[0].lower() == "true",
    )


def send(target: SmtpTarget, password: str | None, message: EmailMessage) -> None:
    context = ssl.create_default_context()
    if target.use_ssl:
        client = smtplib.SMTP_SSL(target.host, target.port, timeout=TIMEOUT_S, context=context)
    else:
        client = smtplib.SMTP(target.host, target.port, timeout=TIMEOUT_S)

    with client:
        if target.starttls:
            client.starttls(context=context)
        if target.user and password:
            client.login(target.user, password)
        client.send_message(message)
